arbnode/espresso_caff_node.go

```diff
@@ -0,0 +1,553 @@
+
+	} else {
+		// If we haven't configured a safety strategy, every delayed message is valid to include in the nodes state.
+		return true, nil
+	}
+
+	// safeBlockNumber will be popluated from here onwards. Any code paths that don't set the variable, return from the function before they get here.
+	delayCount, err := n.delayedMessageFetcher.getDelayedMessageCountAtBlock(safeBlockNumber)
+	if err != nil {
+		log.Warn("Error getting the delayed message count while checking the delayed messages safety tolerance", "err", err)
+		return false, err
+	}
+	if (message.MessageWithMeta.Message.Header.BlockNumber <= safeBlockNumber) && (message.MessageWithMeta.DelayedMessagesRead <= delayCount) {
+		return true, nil
+	}
+	return false, nil
+
+}
+
+// nextMessage wraps the espressoStreamer.Next() method, to handle producing delayed messages by checking they are within the nodes safety tolerance.
+// Returns:
+//   - MessageWithMetadataAndPos: A message, delayed or normally sequenced, that is for the next position in the chain.
+//   - error: If any error is encountered during this function it is propegated to the caller.
+//
+// Semantics:
+//
+//	This function will either produce a message, or an error. When an error is produced, the messageWithMetadataAndPos will be nil.
+//	If the message is populated, the error will be nil.
+func (n *EspressoCaffNode) nextMessage() (*espressostreamer.MessageWithMetadataAndPos, error) {
+	messageWithMetadataAndPos, err := n.espressoStreamer.Next()
+	if err != nil {
+		return nil, err
+	}
+
+	if messageWithMetadataAndPos == nil {
+		return nil, nil
+	}
+
+	if messageWithMetadataAndPos.MessageWithMeta.DelayedMessagesRead == n.nextDelayedCount+1 {
+		log.Info("Getting delayed message", "nextDelayedCount", n.nextDelayedCount)
+		// If this is delayed message, we need to get the message from L1
+		// and replace the message in the messageWithMetadataAndPos
+		message, err := n.delayedMessageFetcher.getDelayedMessage(n.nextDelayedCount)
+		if err != nil {
+			log.Error("failed to get delayed message", "err", err)
+			n.reset(messageWithMetadataAndPos)
+			return nil, err
+		}
+		messageWithMetadataAndPos.MessageWithMeta.Message = message
+		isDelayedMessageWithinSafetyTolerance, err := n.isDelayedMessageWithinSafetyTolerance(messageWithMetadataAndPos)
+		if err != nil {
+			n.reset(messageWithMetadataAndPos)
+			return nil, err
+		}
+		if !isDelayedMessageWithinSafetyTolerance {
+			n.reset(messageWithMetadataAndPos)
+			return nil, fmt.Errorf("Delayed message was not within safety tolerance parameters. The node needs to wait until it is.")
+		}
+		n.nextDelayedCount++
+	}
+	return messageWithMetadataAndPos, nil
+}
+
+func (n *EspressoCaffNode) reset(messageWithMetadataAndPos *espressostreamer.MessageWithMetadataAndPos) {
+	n.espressoStreamer.Reset(messageWithMetadataAndPos.Pos, messageWithMetadataAndPos.HotshotHeight)
+	n.nextDelayedCount = messageWithMetadataAndPos.MessageWithMeta.DelayedMessagesRead
```

Comment:
I'm a bit worried about this check.
Say for example we have a message that comes in at delayed position 2 and we fail to fetch the message from the `DelayedMesageFetcher`. If we reset to the delayed message position in the message, the next time we look at the same delayed message, we won't interpret it as a delayed message because now `n.nextDelayedMessage == messageWithMetadataAndPos` and it won't trigger the `n.nextDelayedMessage+1` check in `nextMessage()`

Reply:
I dont think that is true because n.nextDelayedMessage would not change? Correct?

Reply:
Currently the code would only ever `reset` in `nextMessage` when `messageWithMetadata.MessageWithMeta.DelayedMessagesRead == n.nextDelayedCount +1`.

This means if we cause it to `reset` we update n.nextDelayedCount to `messageWithMetadata.MessageWithMeta.DelayedMessagesRead`. 

This, effectively, increments `n.nextDelayedCount`. to `n.nextDelayedCount+1`

Reply:
okay this means, we dont even need to reset the nextDelayedCount. We can just remove `n.nextDelayedCount = messageWithMetadataAndPos.MessageWithMeta.DelayedMessagesRead`

Reply:
https://github.com/EspressoSystems/nitro-espresso-integration/pull/537/commits/13c2f138f0bb6b1d40121fb3ed779a0ea188e126